src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -930,7 +931,7 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
             Some(ImmutabilityBlame::LocalDeref(node_id)) => {
                 let let_span = self.tcx.hir.span(node_id);
                 match self.local_binding_mode(node_id) {
-                    hir::BindingMode::BindByRef(..) => {
+                    hir::BindingMode::BindByRef(..) => { // FIXME(tschottdorf): vetted
```

Comment:
I think we should use a distinct type for the values stored in the table -- i.e., `ty::BindingMode::BindByRef` or whatever. Were you considering doing that as part of this PR? It would cover a lot of these "vetted" cases -- i.e., a grep for `hir::BindingMode` would yield far fewer results. (Ideally, we would give them distinct names, too...)

Maybe `hir::BindingAnnotation` and `ty::BindingMode` or something like that? Doesn't seem great; I'm shooting for names that convey that what is in HIR is not the final mode, but just covers what the user actually wrote. I envision:

```rust
pub enum BindingAnnotation {
  // No binding annotation given: this means that the final binding mode
  // will depend on whether we have skipped through a `&` refrence
  // when matching. For example, the `x` in `Some(x)` will have binding
  // mode `None`; if you do `let Some(x) = &Some(22)`, it will ultimately be
  // inferred to be by-reference.
  None,

  // Annotated with `mut x` -- could be either ref or not, similar to `None`.
  Mutable,

  // Annotated as `ref`, like `ref x`
  Ref,

  // Annotated as `ref mut x`.
  RefMut,
}
```